### PR TITLE
Adding kid property to JWT headers

### PR DESCRIPTION
**Pull request by jhivandebenoit**
- Generating kid property on jwt headers is now an option
- The value represented by kid is the SHA-2 hash of the public key
- A flag is available in JWTConfigurationDTO to opt into using kid/x5t

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 1]
```diff
@@ -115,6 +116,57 @@ public static String generateHeader(Certificate publicCert, String signatureAlgo
...
+    /**
+     * Utility method to generate JWT header with public certificate thumbprint for signature verification.
+     *
+     * @param publicCert         - The public certificate which needs to include in the header as thumbprint
+     * @param signatureAlgorithm signature algorithm which needs to include in the header
+     * @param useKid - Specifies whether this function should use kid as a thumbprint or x5t
+     * @throws JWTGeneratorException
+     */
+
+    public static String generateHeader(Certificate publicCert, String signatureAlgorithm,boolean useKid) throws
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:128]:
```suggestion
    public static String generateHeader(Certificate publicCert, String signatureAlgorithm, boolean useKid) throws
```

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 2]
```diff
@@ -115,6 +116,57 @@ public static String generateHeader(Certificate publicCert, String signatureAlgo
...
+            } else {
+                jwtHeader.append("\"x5t\":\"");
+                jwtHeader.append(generateThumbprint("SHA-1",publicCert));
+            }
+            jwtHeader.append("\"}");
+            return jwtHeader.toString();
+        } catch (NoSuchAlgorithmException | CertificateEncodingException | UnsupportedEncodingException e) {
+            throw new JWTGeneratorException("Error in generating public certificate thumbprint", e);
+        }
+    }
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:155]:
```suggestion
    }
    
```

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 3]
```diff
@@ -115,6 +116,57 @@ public static String generateHeader(Certificate publicCert, String signatureAlgo
...
 
+    /**
+     * Utility method to generate JWT header with public certificate thumbprint for signature verification.
+     *
+     * @param publicCert         - The public certificate which needs to include in the header as thumbprint
+     * @param signatureAlgorithm signature algorithm which needs to include in the header
+     * @param useKid - Specifies whether this function should use kid as a thumbprint or x5t
+     * @throws JWTGeneratorException
+     */
+
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:127]:
```suggestion
     */
```

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 4]
```diff
@@ -80,6 +80,7 @@ public static String getJWSCompliantAlgorithmCode(String signatureAlgorithm) {
      * @param signatureAlgorithm signature algorithm which needs to include in the header
      * @throws JWTGeneratorException
      */
+
     public static String generateHeader(Certificate publicCert, String signatureAlgorithm) throws
             JWTGeneratorException {
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:86]: Remove the method content and replace it with `generateHeader(Certificate publicCert, String signatureAlgorithm, false)`

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 5]
```diff
@@ -115,6 +116,57 @@ public static String generateHeader(Certificate publicCert, String signatureAlgo
...
+                jwtHeader.append("\"x5t\":\"");
+                jwtHeader.append(generateThumbprint("SHA-1",publicCert));
+            }
+            jwtHeader.append("\"}");
+            return jwtHeader.toString();
+        } catch (NoSuchAlgorithmException | CertificateEncodingException | UnsupportedEncodingException e) {
+            throw new JWTGeneratorException("Error in generating public certificate thumbprint", e);
+        }
+    }
+    private static String generateThumbprint(String hashType,Certificate publicCert) throws CertificateEncodingException,
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:156]:
```suggestion
    private static String generateThumbprint(String hashType, Certificate publicCert) throws CertificateEncodingException,
```

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java (hunk 4 above)
Comment (jhivandebenoit) [reply components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:86]: +1

**Review by VirajSalaka — DISMISSED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java [hunk 6]
```diff
@@ -80,41 +80,65 @@ public static String getJWSCompliantAlgorithmCode(String signatureAlgorithm) {
...
+            throws CertificateEncodingException,
+            UnsupportedEncodingException, NoSuchAlgorithmException {
+        MessageDigest digestValue;
+        byte[] der = publicCert.getEncoded();
+        digestValue = MessageDigest.getInstance(hashType);
+        digestValue.update(der);
+        byte[] digestInBytes = digestValue.digest();
+        String publicCertThumbprint = hexify(digestInBytes);
+        String base64UrlEncodedThumbPrint;
+        base64UrlEncodedThumbPrint = java.util.Base64.getUrlEncoder()
```
Comment (VirajSalaka) [inline components/apimgt/org.wso2.carbon.apimgt.common.gateway/src/main/java/org/wso2/carbon/apimgt/common/gateway/util/JWTUtil.java:137]: `java.util.Base64.getUrlEncoder().withoutPadding()` would remove the equal symbols

**Pull request closed without merging**